packages/webapp/src/journey-validation/task-list-generator.js

```diff
@@ -0,0 +1,222 @@
+const arrayOfAnyComparator = JSON.stringify([ANY])
+
+const sessionMatches = (part, session) => {
+  const checkSessionMatches = ([k, v]) => {
+    if (JSON.stringify(v) === arrayOfAnyComparator) {
+      return session.get(k)?.length > 0
+    }
+
+    return v === ANY ? session.get(k) !== null && session.get(k) !== undefined : session.get(k) === v
+  }
+
+  return Object.entries(part.sessionDataRequired).every(checkSessionMatches)
+}
+
+const sessionMismatches = (part, session) => {
+  const checkSessionMismatch = ([k, v]) => v === ANY ? false : session.get(k) !== null && session.get(k) !== v
+  return Object.entries(part.sessionDataRequired).some(checkSessionMismatch)
+}
+
+const getJourneyStatuses = (schema, session) => {
+  const statuses = schema.journeyParts.map(journey => {
+    for (const part of journey) {
+      if (!sessionMatches(part, session)) {
+        let valid = true
+
+        if (part.sessionMismatchWillInvalidate && sessionMismatches(part, session)) {
+          valid = false
+        }
+
+        if (part.startUrl === schema.startUrl) {
+          return getReturnObject(STATUSES.NOT_STARTED, schema.startUrl, schema.title, valid)
+        }
+
+        return getReturnObject(STATUSES.IN_PROGRESS, part.startUrl, schema.title, valid)
+      }
+    }
+
+    return getReturnObject(STATUSES.COMPLETE, schema.completeUrl, schema.title, true)
+  })
+
+  return statuses
+}
+
+// We should move to Joi to do the validation in a later iteration
+const check = (schema, session) => {
+  const journeyStatuses = getJourneyStatuses(schema, session)
+
+  // Return a complete journey if there is one
+  const completeJourney = journeyStatuses.find(s => s.status === STATUSES.COMPLETE)
+  if (completeJourney) {
+    return completeJourney
+  }
+
+  // If there is a valid in-progress journey return it
+  const inProgressJourney = journeyStatuses.find(s => s.valid && s.status === STATUSES.IN_PROGRESS)
+  if (inProgressJourney) {
+    return inProgressJourney
+  }
+
+  // Found no complete or valid in-progress journeys, so return not started
+  return getReturnObject(STATUSES.NOT_STARTED, schema.startUrl, schema.title, true)
+}
+
+const getJourneySectionStatus = (journey, section, session) => {
+  return check(journeyDefinition[journey][section], session)
+}
+
+const getTaskList = (journey, session) => {
+  const applicantInfoTask = getJourneySectionStatus(journey, REGISTRATION.APPLICANT_INFO, session)
+  const landOwnershipTask = getJourneySectionStatus(journey, REGISTRATION.LAND_OWNERSHIP, session)
+  const siteBoundaryTask = getJourneySectionStatus(journey, REGISTRATION.SITE_BOUNDARY, session)
+  const habitatInfoTask = getJourneySectionStatus(journey, REGISTRATION.HABITAT_INFO, session)
+  const legalAgreementTask = getJourneySectionStatus(journey, REGISTRATION.LEGAL_AGREEMENT, session)
+  const localLandChargeTask = getJourneySectionStatus(journey, REGISTRATION.LOCAL_LAND_CHARGE, session)
+
```

Comment:
thought in progress -


```suggestion

const taskSectionFactory = (taskTitle, tasks, id) => {
  return {
    taskTitle,
    tasks: tasks.map(task => {
      return {
        title: task.title,
        status: task.status,
        url: task.url,
        id: task.id
      }
    })
  }
}
```

Reply:
I don't mean to add it to at this line obviously :). I just put that here for us to discuss later

Reply:
refactoring tickets to follow